Use Elasticsearch index aliases to avoid search downtime

peterbe opened the pull request:
Part of #3098

(the reason this PR doesn't close issue #3098) is because the `prod-build.yml` still uses `--update`)

peterbe commented:
What's cool about this PR is that it will not require any manual intervention. 
The first time this gets run it will notice that `mdn_docs` was an **index** and not an **alias**. And before setting the alias (`mdn_docs --> mdn_docs_20210331`) it will delete the old `mdn_docs` index.

But then, the next day, it will not have to worry about that stuff. 

Also notice that I set the Stage site to start using this. So the plan is to 

1. Review this PR and land it
2. Let it do its thing on Stage for a day or two
3. Change `prod-build.yml` to enable it on Prod
4. Go back and remove the rather temporary code that deletes `mdn_docs` if it was an index (and not an alias)

Throughout this we will not need to do anything on the Kuma side.

escattone commented on deployer/src/deployer/search/models.py:
```diff
@@ -9,6 +11,12 @@
 
+# Note, this is the name that the Kuma code will use when sending Elasticsearch
+# search queries.
+# We always build and index that is called something based on this name but with
```
```suggestion
# We always build an index that is called something based on this name but with
```

escattone commented on deployer/src/deployer/search/__init__.py:
```diff
@@ -40,42 +40,32 @@ def index(
     if not update:
```
I don't really know what `Document._index` returns, so I'm probably off here, but I assume it returns the index for the day (i.e., `mdn_docs_2021mmdd`)? If we're in `update` mode, it seems that `document_index` will be the new one (based on today's date) and not the previous one that we want to update? Also, if we're **not** in `update` mode, we can't run this more than once a day, or else we'll delete the same index the alias is pointing to, right?

peterbe replied:
No, I don't think that's a concern. 
Suppose that we create a brand new index at 00:00 on mdn_docs_20210401 and we populate it. Then at 01:00 we use `--update` and "reindex" (aka. cake on) another 100 documents, they will update `mdn_docs_20210401` which is fine because Kuma will always search on the alias which, for the rest of the day is pointing to `mdn_docs_20210401`. 

The alias just means you're searching in `mdn_docs` which is always the last index to be populated independent of the day. 

But imagine if we "copied" mdn_docs_20210401 into mdn_docs, then mdn_docs would be missing the new additions if you go back to populate more into mdn_docs_20210401. But aliases aren't copies. They're just aliases. 

Lastly...

> if we're not in update mode, we can't run this more than once a day, or else we'll delete the same index the alias is pointing to?

Yes, it's the truth. If you run `poetry run deployer search-index /path/to/build` TWICE in one day, it will create a downtime window. It's almost as if we're saying "I change my mind. I want to start over for today's index. Delete the old one and fill it all back up again."
But this is not going to happen outside of local development. Once we start adding partial builds throughout the day, the timeline will be:

- Monday morning: `poetry run deployer search-index /path/to/build` (will delete and populate a new index)
- Monday lunch: `poetry run deployer search-index /path/to/build --update` (will append to Monday's index)
- Monday afternoon: `poetry run deployer search-index /path/to/build --update` (will append to Monday's index)
- Tuesday morning: `poetry run deployer search-index /path/to/build` (will delete and populate a new index)
- Tuesday lunch: `poetry run deployer search-index /path/to/build --update` (will append to Tuesday's index)
- Tuesday afternoon: `poetry run deployer search-index /path/to/build --update` (will append to Tuesday's index)

We currently have `prod-build.yml` which we execute every 24h. 
But hopefully, we'll (soon) have `prod-partial-build.yml` which builds based on `git` history several times a day but then it'll use `poetry run deployer search-index /path/to/build --update` on those.

escattone replied:
Assuming `Document._index` returns the index for the day (i.e., `mdn_docs_2021mmdd`) I still think this logic is confusing (i.e., there are hidden assumptions that make it fragile). It seems to me that if `--update` is specified, we want to ask the alias for its current index, and use that one, instead of calling `Document._index`. In other words, as the code is currently written, it seems that `--update` only works if you do it on the same day that you already created an index. So, if you create a new index on Monday, don't do anything else until Tuesday, and on Tuesday only try to do an update, it will fail. I guess I'm saying that `--update` feels fragile because we're not asking the alias for its current index and using that, which seems to me what we should do when `--update` is specified.

Also, for the case when we're not using `--update`, we're assuming that we're only going to do that once per day. If we do that more than once, we'll have downtime. Why not use the time of day (to the minute or even second) as part of the index name? That way we can run a full reindex multiple times a day (if we want) with no downtime.

I guess I'm saying that the current implementation assumes a daily usage pattern (only one full reindex per day, and updates only work for that day) that isn't clear from the CLI options, and with some minor adjustments, we can eliminate those assumptions.

peterbe commented on deployer/src/deployer/search/__init__.py:
```diff
@@ -40,42 +40,32 @@ def index(
-                    skipped.append(1)
+        root = Path(buildroot)
+        for doc in walk(root):
+            if doc in already:
```
You're right! Good catch

peterbe commented on deployer/README.md:
```diff
@@ -131,6 +131,36 @@ This is to make it convenient in GitHub Actions to control the
+- Today:
+  - create index `mdn_docs_20210331`
+  - populate `mdn_docs_20210331` (could take a long time)
+  - remove alias `mdn_docs`
+  - put new alias `mdn_docs --> mdn_docs_20210331`
```
No, I don't think they're a "single atomic operation". If, god forbid, the power was to go out between the `remove_alias` and the `put_alias` we'd be screwed. But it's the only way I could figure it out to work. 

I actually tested this. If I commented out the `remove_alias()` step what would happen is that `mdn_docs --> mdn_docs_YESTERDAY AND mdn_docs_TODAY` and when you make a search through Kuma you get twice as many matches. 

What would be ideal is to use some sort of atomic operation like you can do with a  proper relational database. 

Another alternative would be to rearrange the order so it's 
```
  - put new alias `mdn_docs --> mdn_docs_TODAY`
  - remove alias `mdn_docs --> mdn_docs_YESTERDAY`
 ```
then, if the power was to go out in between those two operations you'd just have duplicates of every search result. 

Perhaps I'm wrong! Perhaps there is a fancy Elasticsearch operation where you bundle them into 1 single POST request. Want me to look deeper into that? I didn't see anything in the Python API but it looks like mayhaps it's possible with the raw JSON API.

escattone replied:
Oh, I see, but I think the underlying `Elasticsearch` client that `elasticsearch_dsl` uses supports it via `client.indices.update_aliases()` (see https://elasticsearch-py.readthedocs.io/en/v7.12.0/api.html#elasticsearch.client.IndicesClient.update_aliases and https://www.elastic.co/guide/en/elasticsearch/reference/7.12/indices-aliases.html#indices-aliases-api-rename-alias-ex) but it doesn't seems that `elasticsearch_dsl` exposes that?

escattone replied:
Hm, I wonder if you pass the `body` keyword argument to `put_aliases()` (using the `remove` and `add` `actions`) it would work? It seems so strange that this isn't easier to do using `elasticsearch_dsl` as it seems that it would be a super common need.

escattone replied:
Hm, I'm reading the `elasticsearch_dsl` source code, and it might work to use:
```python
document_index.connection.update_aliases(body)
```
where I guess `body` is the underlying JSON body as described in https://www.elastic.co/guide/en/elasticsearch/reference/7.12/indices-aliases.html#indices-aliases-api-request-body ?

escattone replied:
I think we should look some more into doing this atomically. From my reading of the ElasticSearch API, it's possible. In addition to what I found in my comments above, see the section `Reindexing with no downtime` in https://www.objectrocket.com/blog/elasticsearch/elasticsearch-aliases/.

peterbe commented:
In conclusion:

1. Use an index name which is down to the second so you *can* deploy a full (re)build any and multiple times a day if you have to. 
2. Make the `--update` operate based on "reverse lookup" the alias. E.g. "I'm about to do an --update, what's the current index name that the `mdn_docs` alias is currently serving?"
3. Combine the put and remove alias all in one single POST

escattone commented:
@peterbe Yes, great summary.

peterbe commented:
@escattone Ok. I've carefully tested it now. 
Once this is merged, we don't need to change anything. The old way of doing things in Prod will continue to work and the first time it runs on Stage it's going to delete the old (un-aliased) index and apply the new functionality. 
r?

escattone commented on deployer/src/deployer/search/__init__.py:
```diff
@@ -122,6 +129,57 @@ def get_progressbar():
+
+        connection.indices.update_aliases({"actions": alias_updates})
+        click.echo(
+            f"Put the {INDEX_ALIAS_NAME!r} alias from old index "
+            f"to point to {document_index._name}"
```
```suggestion
            f"Reassign the {INDEX_ALIAS_NAME!r} alias from old index "
            f"to {document_index._name}"
```

escattone reviewed (changes requested):
@peterbe Just some nits, mostly in the comments.

peterbe commented:
@escattone I addressed all the feedback. And I boldly made it so that `prod-build.yml` and `stage-build.yml` both stop the `--update`. Because I figured, this new technique is much more harmless. In the old way, it would delete the entire index before it starts (to re-create it) and the fill it up. So, if the bulk publishing is still slow (e.g. 40min) it doesn't matter because it doesn't re-arrange the aliases until *after* it has all worked.

escattone reviewed (approved):
@peterbe Thanks for all of your work on this! It's a really nice step forward!
